lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,117 @@
+    /// \param [in] hook The function to run when the json path matches.
+    ///
+    /// \b Example
+    /// \code{.cpp}
+    /// namespace cr = irods::experimental::configuration_reload;
+    /// cr::add_hook({cr::CATCHALL_KEYWORD}, [&](const std::string& json_path,
+    ///                                          const nlohmann::json& json_patch_op)
+    /// {
+    ///    if(json_path.substr(0,29) != "/example_configuration_object") {
+    ///         return;
+    ///    }
+    ///    handle_update(json_patch_op);
+    /// });
+    /// cr::add_hook({"/example_configuration_object/circumference_to_diameter_ratio"},
+    ///              [](const std::string& json_path, const nlohmann::json& json_patch_op){
+    ///    // Update the "PI" global value in the static lua state.
+    ///    if(json_patch_op.at("op")=="insert"||json_patch_op.at("op")=="replace"){
+    ///        lua_pushnumber( Lua_state, json_patch_op.at("value").get<double>());
+    ///        lua_setglobal( Lua_state, "PI");
+    ///    }
+    /// });
+    /// \endcode
+    void add_hook(const auto& hook_triggers, hook&& hook)
+    {
+        size_t id = hooks.size();
+        // Take ownership, we don't need to deal with any missing closures.
+        // those can be nasty to debug.
+        hooks.emplace_back(std::move(hook));
+        for (const auto& name : hook_triggers) {
+            edge_sets.insert(std::make_pair(name, id));
+        }
+    }
+
+    /// \brief Run the hooks, the changes are a list of json patches.
+    /// \param [in] changes The list of changes in json patch format, see
+    /// RFC6902 for more details about that format.
+    /// \code{.cpp}
+    /// namespace cr = irods::experimental::configuration_reload;
+    /// cr::add_hook({cr::CATCHALL_KEYWORD}[](const std::string& json_path, const nlohmann::json& patch){
+    ///    std::cout<<patch<<std::endl;
+    /// });
+    /// \endcode
+    /// Change the delay_server_sleep_time_in_seconds in the configuration and send a
+    /// SIGHUP signal to the toplevel irods process. The server should write
+    /// something like
+    /// `{ "path":"/advanced_settings/delay_server_sleep_time_in_seconds",
+    ///    "op":"replace", "value":<something> }` to stdout.
+    void run_hooks(const nlohmann::json& changes)
+    {
+        // The *CATCHALL* hook is triggered on every change in configuration value.
+        auto catch_all_hooks = edge_sets.equal_range(CATCHALL_KEYWORD);
+
+        for (const auto& operation : changes) {
+            const std::string& name = operation.at("path").get_ref<const std::string&>();
+
+            for (auto it = catch_all_hooks.first; it != catch_all_hooks.second; ++it) {
+                hooks[it->second](name, operation);
+            }
+
+            auto range = edge_sets.equal_range(name);
+            for (auto it = range.first; it != range.second; ++it) {
+                hooks[it->second](name, operation);
+            }
+        }
+    }
```

Comment:
The definition of this function needs to move to the cpp file to avoid ODR violations